lib/logstash/outputs/mongodb.rb

```diff
@@ -67,22 +70,32 @@ def register
             end
           end
         end
+
+        break if @closed.true?
       end
     end
-  end # def register
+  end
 
   def receive(event)
     begin
       # Our timestamp object now has a to_bson method, using it here
       # {}.merge(other) so we don't taint the event hash innards
       document = {}.merge(event.to_hash)
+
       if !@isodate
-        # not using timestamp.to_bson
-        document["@timestamp"] = event.timestamp.to_json
+        timestamp = event.timestamp
+        if timestamp
+          # not using timestamp.to_bson
+          document["@timestamp"] = timestamp.to_json
+        else
+          @logger.warn("Cannot set MongoDB document `@timestamp` field because it does not exist in the event", :event => event)
+        end
       end
+
       if @generateId
-        document["_id"] = BSON::ObjectId.new(nil, event.timestamp)
+        document["_id"] = BSON::ObjectId.new
       end
+
       if @bulk
         @@mutex.synchronize do
           collection = event.sprintf(@collection)
```

Comment:
nitpick unrelated to your proposed changes: we should pull this sprintf out of the global mutex block to improve performance.

Reply:
sure - but same comment about the concurrency code.

Reply:
done.